shipyard_blueprints/snowflake/shipyard_snowflake/snowpark.py

```diff
@@ -0,0 +1,124 @@
+import os
+import snowflake.snowpark.functions as F
+import snowflake.snowpark.types as T
+import pandas as pd
+from snowflake.snowpark.session import Session
+from snowflake.snowpark.window import Window
+from shipyard_templates import ExitCodeException, Database
+
+
+class SnowparkClient(Database):
+    def __init__(
+        self,
+        username,
+        pwd,
+        database=None,
+        account=None,
+        warehouse=None,
+        schema=None,
+        role=None,
+    ):
+        self.username = username
+        self.pwd = pwd
+        self.account = account
+        self.database = database
+        self.warehouse = warehouse
+        self.schema = schema
+        self.role = role
+
+        super().__init__(
+            username,
+            pwd,
+            account=account,
+            warehouse=warehouse,
+            schema=schema,
+            database=database,
+        )
+
+    def connect(self):
+        params = {
+            "account": self.account,
+            "user": self.username,
+            "password": self.pwd,
+            "warehouse": self.warehouse,
+            "database": self.database,
+            "schema": self.schema,
+            "role": self.role,
+        }
+
+        try:
+            session = Session.builder.configs(params).create()
+        except Exception as e:
+            self.logger.error(f"Error from snowpark {e}")
+            raise ExitCodeException(
+                message="Could not establish Snowpark Session",
+                exit_code=self.EXIT_CODE_INVALID_CREDENTIALS,
+            )
+        else:
+            return session
+
+    def upload(
+        self, session: Session, df: pd.DataFrame, table_name: str, overwrite=True
+    ):
+        """
+        Uploads a pandas dataframe to a table in Snowflake using the Snowpark API
+
+        Args:
+            overwrite (): Whether to overwrite the existing table. True to replace, False to append to table (if it exists)
+            session: The Snowpark Session established by the connect method
+            df: The pandas dataframe to load
+            table_name: The name of the Snowflake Table to write to
+        """
+
+        self.logger.info("Attempting to write pandas dataframe to Snowflake")
+        try:
+            session.write_pandas(
+                df, table_name=table_name, auto_create_table=True, overwrite=overwrite
+            )
+            self.logger.info(f"Successfully wrote data to {table_name}")
+
+        except Exception as e:
```

Comment:
At a glance this error may lead to a misleading exit code. For example if the user attempts to call this method with invalid credentials would that error be raised before this exception is caught? It may say invalid credentials with an exit code `EXIT_CODE_INVALID_UPLOAD_VALUE`

Reply:
You would need to actually connect first before running this, so credential errors are actually not possible as those would be caught in the `connect` method

Reply:
the invalid credentials was an example of another error that could occur. Ideally the broad exception catch should be UNKNOWN ERROR. If there isn't a more specific error code/msg or exception thrown it may not be possible and invalid upload may still be the best option.

in a perfect world the flow would be more along the lines of
```
try:
    code
except ValueError as error:
     raise .. EXIT_CODE_INVALID_UPLOAD_VALUE
 except exception as error
     raise UNKNOWN ERROR
```
this comment is food for thought